### ISSUE TITLE
Argument checks in exported functions

### ISSUE THREAD

**Issue by tdeenes** [ManuelHentschel/vscDebugger#57]
In https://github.com/ManuelHentschel/vscDebugger/blob/0039991449ce90483d69c65a126c7362628543ae/R/customVarInfo.R#L223 I've tried to add a kind of check for `varInfo`s. However, my idea was to add a function like this to `.vsc.addVarInfo`, so that it's not possible (or at least more difficult) for users to add invalid varInfos. Would it be possible to reuse this for testing? E.g. check all entries in `defaultVarinfos` at build time and use the same code to test new entries at runtime?

_Originally posted by @ManuelHentschel in https://github.com/ManuelHentschel/vscDebugger/pull/56#issuecomment-643530202_

**Comment by tdeenes** [comment]
I already wanted to ask this, but @ManuelHentschel raised a related question (see above) recently, so let it discuss here.

Normally I do the following:

1. In all exported functions, I perform assertions on all input arguments. I use the **checkmate** package for this purpose. Fast, very stable, and actively maintained (disclaimer: I am a contributor).

1. If a full argument check is too expensive (e.g., the given function is called many-many times), I either add an extra argument (e.g., `check_arg = TRUE` by default, which can be set to `check_arg = FALSE`), or write a "pure" (internal) function without the checks, and call that internal function in the exported function, but also export the internal function for "power user use cases". 

 
1. In unit tests, I usually test that the assertions are in place and catch falsy arguments.

1. If I have function or object factories, I try to use them for internal purposes as well.

So I would rather implement `.vsc.addVarInfo` in such a way, that it performs assertions on its input arguments. Then, `defaultVarInfo` could be the result of a series of `.vsc.addVarInfo` calls with the required arguments -> this way we ensure that we do not create erroneous definitions.

**Comment by ManuelHentschel** [comment]
The **checkmate** package looks very nice! I don't think performance is a big issue for us. Maybe if the user tries to view the entire content of the base-namespace after each step, but I guess that's not a frequent usecase...

I've done some restructuring lately and started documenting the most important functions/arguments/returns in typescript declaration files (https://github.com/ManuelHentschel/vscDebugger/tree/setVar/d.ts). I was constantly checking the [DAP](https://github.com/ManuelHentschel/vscDebugger/blob/setVar/d.ts/debugProtocol.d.ts) anyways, so I just added some placeholder  [R types](https://github.com/ManuelHentschel/vscDebugger/blob/setVar/d.ts/RTypes.d.ts) to document the modifications used in the R code as well. These .d.ts files should be fairly straightforward to implement as unit tests.

Since the messages to and from the debug adapter are represented as lists in R, I am using an `lget()` function to retrieve list items without worrying about gettings `NULL`s all the time: https://github.com/ManuelHentschel/vscDebugger/blob/b6909d4d8044c454b9075e27990676744194abb6/R/utils.R#L116

Maybe it makes sense to include a type assertion here as well, e.g. `lget(list, entry, default=NULL, assertType="string")` etc.

Adding the `defaultVarInfo` during runtime sounds like a very good idea. I've wasted way too much time finding errors in the list that were not caught during build time :D

**Comment by tdeenes** [comment]
I am pretty busy this week, but will try to come back to you early next week.

**Comment by tdeenes** [comment]
> Adding the `defaultVarInfo` during runtime sounds like a very good idea.

Actually I did not suggest to add `defaultVarInfo` during runtime. What I was suggested was that it should be created by the `.vsc.addVarInfo` function, during build time. With pseudo-code:
```r
#' Add custom rules to display variable information during debugging
#'
#' @export
.vsc.addVarInfo <- function(...) {
  ...
}

# For contributors: this is where the default var infos are defined;
# make sure you do not try to circumvent .vsc.addVarInfo() because 
# it performs thorough argument checks
session$defaultVarInfo <- list(
  ## NULLs
  .vsc.addVarInfo(
      name = 'NULL',
      doesApply = is.null,
      childVars = list(),
      hasChildren = FALSE,
      shortType = '',
      longType = 'NULL',
      toString = 'NULL'
    ),
   ## active bindings
  .vsc.addVarInfo(
      name = 'ActiveBinding',
      doesApply = function(v) inherits(v, '.vsc.activeBinding'),
      longType = 'active binding',
      hasChildren = TRUE,
      toString = 'Active binding',
      includeAttributes = FALSE, # TODO: Can active bindings have non-active attributes?
      childVars = list(),
      internalAttributes = function(v) {
        list(
          names = list('bindingFunction'),
          values = list(v$bindingFunction)
        )
      }
    )
)
```

**Comment by tdeenes** [comment]
Or even better, `list` should be replaced by `createDefaultVarInfos`, where:

```r
createDefaultVarInfos <- function(...) {
  structure(list(...), class = "some_class")
}
```

**Comment by ManuelHentschel** [comment]
> 
> Actually I did not suggest to add `defaultVarInfo` during runtime. What I was suggested was that it should be created by the `.vsc.addVarInfo` function, during build time. With pseudo-code:

You're right, that's probably the best way to do it. I got a little confused with the difference between a hard-coded list and a list produced at build-time, my bad.
